From e0d799858782a3dede104e5f24544b5b21ea6422 Mon Sep 17 00:00:00 2001
From: "Richard W.M. Jones" <rjones@redhat.com>
Date: Wed, 8 Apr 2026 15:38:39 +0100
Subject: [PATCH] daemon: xfs: Rewrite xfs_repair implementation in OCaml

Rewrite the implementation of xfs_repair from C into OCaml.  No
functional change in this version.

(cherry picked from commit 29e1b15cfd779bd2ac4e655357a2daa883592f9e)
---
 daemon/xfs.c              | 115 --------------------------------------
 daemon/xfs.ml             |  60 ++++++++++++++++++++
 generator/actions_core.ml |   1 +
 3 files changed, 61 insertions(+), 115 deletions(-)

diff --git a/daemon/xfs.c b/daemon/xfs.c
index 06e5fc072..ea5303d36 100644
--- a/daemon/xfs.c
+++ b/daemon/xfs.c
@@ -230,121 +230,6 @@ do_xfs_admin (const char *device,
   return 0;
 }
 
-int
-do_xfs_repair (const char *device,
-               int forcelogzero, int nomodify,
-               int noprefetch, int forcegeometry,
-               int64_t maxmem, int64_t ihashsize,
-               int64_t bhashsize, int64_t agstride,
-               const char *logdev, const char *rtdev)
-{
-  int r;
-  CLEANUP_FREE char *err = NULL, *buf = NULL;
-  const char *argv[MAX_ARGS];
-  char maxmem_s[64];
-  char ihashsize_s[70];
-  char bhashsize_s[70];
-  char agstride_s[74];
-  size_t i = 0;
-  int is_device;
-
-  ADD_ARG (argv, i, "xfs_repair");
-
-  /* Optional arguments */
-  if (optargs_bitmask & GUESTFS_XFS_REPAIR_FORCELOGZERO_BITMASK) {
-    if (forcelogzero)
-      ADD_ARG (argv, i, "-L");
-  }
-  if (optargs_bitmask & GUESTFS_XFS_REPAIR_NOMODIFY_BITMASK) {
-    if (nomodify)
-      ADD_ARG (argv, i, "-n");
-  }
-  if (optargs_bitmask & GUESTFS_XFS_REPAIR_NOPREFETCH_BITMASK) {
-    if (noprefetch)
-      ADD_ARG (argv, i, "-P");
-  }
-  if (optargs_bitmask & GUESTFS_XFS_REPAIR_FORCEGEOMETRY_BITMASK) {
-    if (forcegeometry) {
-      ADD_ARG (argv, i, "-o");
-      ADD_ARG (argv, i, "force_geometry");
-    }
-  }
-
-  if (optargs_bitmask & GUESTFS_XFS_REPAIR_MAXMEM_BITMASK) {
-    if (maxmem < 0) {
-      reply_with_error ("maxmem must be >= 0");
-      return -1;
-    }
-    snprintf (maxmem_s, sizeof maxmem_s, "%" PRIi64, maxmem);
-    ADD_ARG (argv, i, "-m");
-    ADD_ARG (argv, i, maxmem_s);
-  }
-
-  if (optargs_bitmask & GUESTFS_XFS_REPAIR_IHASHSIZE_BITMASK) {
-    if (ihashsize < 0) {
-      reply_with_error ("ihashsize must be >= 0");
-      return -1;
-    }
-    snprintf (ihashsize_s, sizeof ihashsize_s, "ihash=" "%" PRIi64, ihashsize);
-    ADD_ARG (argv, i, "-o");
-    ADD_ARG (argv, i, ihashsize_s);
-  }
-
-  if (optargs_bitmask & GUESTFS_XFS_REPAIR_BHASHSIZE_BITMASK) {
-    if (bhashsize < 0) {
-      reply_with_error ("bhashsize must be >= 0");
-      return -1;
-    }
-    snprintf (bhashsize_s, sizeof bhashsize_s, "bhash=" "%" PRIi64, bhashsize);
-    ADD_ARG (argv, i, "-o");
-    ADD_ARG (argv, i, bhashsize_s);
-  }
-
-  if (optargs_bitmask & GUESTFS_XFS_REPAIR_AGSTRIDE_BITMASK) {
-    if (agstride < 0) {
-      reply_with_error ("agstride must be >= 0");
-      return -1;
-    }
-    snprintf (agstride_s, sizeof agstride_s, "ag_stride=" "%" PRIi64, agstride);
-    ADD_ARG (argv, i, "-o");
-    ADD_ARG (argv, i, agstride_s);
-  }
-
-
-  if (optargs_bitmask & GUESTFS_XFS_REPAIR_LOGDEV_BITMASK) {
-    ADD_ARG (argv, i, "-l");
-    ADD_ARG (argv, i, logdev);
-  }
-
-  if (optargs_bitmask & GUESTFS_XFS_REPAIR_RTDEV_BITMASK) {
-    ADD_ARG (argv, i, "-r");
-    ADD_ARG (argv, i, rtdev);
-  }
-
-  is_device = is_device_parameter (device);
-  if (!is_device) {
-    buf = sysroot_path (device);
-    if (buf == NULL) {
-      reply_with_perror ("malloc");
-      return -1;
-    }
-    ADD_ARG (argv, i, "-f");
-    ADD_ARG (argv, i, buf);
-  } else {
-    ADD_ARG (argv, i, device);
-  }
-
-  ADD_ARG (argv, i, NULL);
-
-  r = commandrv (NULL, &err, argv);
-  if (r == -1) {
-    reply_with_error ("%s: %s", device, err);
-    return -1;
-  }
-
-  return r;
-}
-
 int64_t
 xfs_minimum_size (const char *path)
 {
diff --git a/daemon/xfs.ml b/daemon/xfs.ml
index 157a84058..592300c98 100644
--- a/daemon/xfs.ml
+++ b/daemon/xfs.ml
@@ -171,3 +171,63 @@ let xfs_info dev =
     xfs_rtblocks     = find "realtime.blocks"    Int64.of_string;
     xfs_rtextents    = find "realtime.rtextents" Int64.of_string;
   }
+
+let xfs_repair
+      ?(forcelogzero = false)
+      ?(nomodify = false)
+      ?(noprefetch = false)
+      ?(forcegeometry = false)
+      ?maxmem ?ihashsize ?bhashsize ?agstride ?logdev ?rtdev device =
+  let args = ref [] in
+  if forcelogzero then List.push_back args "-L";
+  if nomodify then List.push_back args "-n";
+  if noprefetch then List.push_back args "-P";
+  if forcegeometry then List.push_back_list args [ "-o"; "force_geometry" ];
+  (match maxmem with
+   | None -> ()
+   | Some m when m < 0_L -> invalid_arg "maxmem must be >= 0"
+   | Some m ->
+      List.push_back_list args [ "-m"; Int64.to_string m ]
+  );
+  (match ihashsize with
+   | None -> ()
+   | Some i when i < 0_L -> invalid_arg "ihashsize must be >= 0"
+   | Some i ->
+      List.push_back_list args [ "-o"; sprintf "ihash=%Ld" i ]
+  );
+  (match bhashsize with
+   | None -> ()
+   | Some i when i < 0_L -> invalid_arg "bhashsize must be >= 0"
+   | Some i ->
+      List.push_back_list args [ "-o"; sprintf "bhash=%Ld" i ]
+  );
+  (match agstride with
+   | None -> ()
+   | Some i when i < 0_L -> invalid_arg "agstride must be >= 0"
+   | Some i ->
+      List.push_back_list args [ "-o"; sprintf "ag_stride=%Ld" i ]
+  );
+  (match logdev with
+   | None -> ()
+   | Some d -> List.push_back_list args [ "-l"; d ]
+  );
+  (match rtdev with
+   | None -> ()
+   | Some d -> List.push_back_list args [ "-r"; d ]
+  );
+
+  if is_device_parameter device then (
+    List.push_back args device
+  )
+  else (
+    (* It's a filesystem image stored in the mounted filesystem.
+     * This slightly dubious usage has been present since the
+     * API was first added to libguestfs back in 2012 (commit 7036a3bccf).
+     *)
+    List.push_back args "-f";
+    List.push_back args (Sysroot.sysroot_path device)
+  );
+
+  (* Run the xfs_repair command, returning the status directly. *)
+  let r, _, _ = commandr "xfs_repair" !args in
+  r
diff --git a/generator/actions_core.ml b/generator/actions_core.ml
index 5876188e9..5ded987f4 100644
--- a/generator/actions_core.ml
+++ b/generator/actions_core.ml
@@ -7457,6 +7457,7 @@ the lazy-counters setting (ie. C<lazycounter> parameter has no effect).|} };
 
   { defaults with
     name = "xfs_repair"; added = (1, 19, 36);
+    impl = OCaml "Xfs.xfs_repair";
     style = RInt "status", [String (Dev_or_Path, "device")], [OBool "forcelogzero"; OBool "nomodify"; OBool "noprefetch"; OBool "forcegeometry"; OInt64 "maxmem"; OInt64 "ihashsize"; OInt64 "bhashsize"; OInt64 "agstride"; OString "logdev"; OString "rtdev"];
     optional = Some "xfs";
     tests = [
-- 
2.47.3

